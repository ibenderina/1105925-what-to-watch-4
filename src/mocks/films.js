import {chooseRandomInt} from "../utils/utils";

export default [
  {
    id: 0,
    title: `Dead in a Week (Or Your Money Back)`,
    src: `https://pets.mail.ru/sharepic/68/10401.png`,
    description: `не выходи из офиса, не находи предлог,
                  зачем тебе солнце, если ты куришь «Vogue»?
                  снаружи всё злостно-хитро поэтому тут – уют,
                  не выходи из офиса, на улице точно убьют.
                  о, не выходи из офиса, не спускай взгляд с монитора,
                  потому что монитор тебе как хасиду тора,
                  и также как она абсолютно непререкаем,
                  не выходи из офиса, Авель, за дверью тебя ждёт Каин,
                  не выходи из офиса, займись квартальным отчётом,
                  если зам позвонит, не колеблясь назови идиотом,
                  зачем выходить оттуда, куда ты утром вернёшься,
                  не выходи из офиса, в аптечке имеется «Нош-па».
                  о, не выходи из офиса, отстукивай клавиатурой марши,
                  не думай о Юле, о Жанне, об Ире, о Маше;
                  с проходной веет мартом и охранника «Дошираком»,
                  не выходи из офиса, стань отшельником, но не раком.
                  не выходи из офиса, о, пускай только офис
                  знает твой истинный облик, приправленный кофе с
                  корицей, сливками, прочим; поёт мп3-соловей,
                  не выходи из офиса, на улице, чай, не Кувейт.
                  не подчиняйся общим укладам, отправь их в топку,
                  зачем по утрам познавать метро или пробку,
                  не выходи из офиса, прими шум принтера за овации,
                  спрячься от эмиссии, кризиса, безработицы, девальвации.`,
    director: `Dude`,
    starring: `Suicidal Guy`,
    genre: `Drama`,
    date: `One Million years BC`,
    background: `https://static.irk.ru/media/img/site/gallery/30414/58476b75-5f6f-42fb-97ff-9b97f79124cd_jpg_730x1000_q85.jpg`,
    ratingScore: chooseRandomInt(0, 10),
    ratingCount: chooseRandomInt(1, 450),
  }, {
    id: 1,
    title: `The Gentlemen`,
    src: `https://twizz.ru/wp-content/uploads/-000//1/animals-about-to-drop-album-photos-58aeb0ce59ff1__700.jpg`,
    description: `Мой интерес, бесспорно, околофутбол. Когда играли двор на двор, я пацанам носил мячи.`,
    director: `Dude`,
    starring: `The Gentlemen`,
    genre: `Comedy`,
    date: `One Million years AC`,
    background: `https://99px.ru/sstorage/53/2017/12/tmb_217309_894087.jpg`,
    ratingScore: chooseRandomInt(0, 10),
    ratingCount: chooseRandomInt(1, 450),
  }, {
    id: 2,
    title: `Lion`,
    src: `https://medialeaks.ru/wp-content/uploads/2019/08/glavnaya-1.jpg`,
    description: `Норм фильм.`,
    director: `Dude`,
    starring: `Hot Dev Patel`,
    genre: `Drama`,
    date: `A long, long time ago`,
    background: `https://img2.goodfon.ru/wallpaper/nbig/6/68/minimalizm-oblaka-blue.jpg`,
    ratingScore: chooseRandomInt(0, 10),
    ratingCount: chooseRandomInt(1, 450),
  }, {
    id: 3,
    title: `Chappie`,
    src: `https://www.thepaws.net/wp-content/uploads/2018/10/funny-chihuahua.jpg`,
    description: `La-la-la`,
    director: `Dude`,
    starring: `Cute Robot`,
    genre: `Comedy`,
    date: `2015`,
    background: `https://img.desktopwallpapers.ru/flowers/pics/wide/1920x1200/fd796b086cb82fea063cd6ecafb73caf.jpg`,
    ratingScore: chooseRandomInt(0, 10),
    ratingCount: chooseRandomInt(1, 450),
  }, {
    id: 4,
    title: `Guns Akimbo`,
    src: `https://memepedia.ru/wp-content/uploads/2018/12/hamster.jpg`,
    description: `Bla-bla-bla`,
    director: `Dude`,
    starring: `Harry Potter`,
    genre: `Thriller`,
    date: `Once upon a time`,
    background: `https://img5.goodfon.ru/wallpaper/big/6/80/jackson-lake-mount-moran-grand-teton-lake-ozero-gory.jpg`,
    ratingScore: chooseRandomInt(0, 10),
    ratingCount: chooseRandomInt(1, 450),
  }, {
    id: 5,
    title: `Undisputed`,
    src: `https://stihi.ru/pics/2013/04/04/1852.jpg`,
    description: `Hare Krishna, everybody`,
    director: `Dude`,
    starring: `Hot fighting men`,
    genre: `Thriller`,
    date: `Not today`,
    background: `https://w-dog.ru/wallpapers/9/19/448077089475704.jpg`,
    ratingScore: chooseRandomInt(0, 10),
    ratingCount: chooseRandomInt(1, 450),
  }, {
    id: 6,
    title: `En man som heter Ove`,
    src: `https://techtoday.in.ua/wp-content/uploads/2018/01/GRUMPYCAT.jpg`,
    description: `Hello`,
    director: `Dude`,
    starring: `Grouchy old man`,
    genre: `Drama`,
    date: `Recently`,
    background: `https://linchakin.com/files/word/1000/411/1.jpg`,
    ratingScore: chooseRandomInt(0, 10),
    ratingCount: chooseRandomInt(1, 450),
  }, {
    id: 7,
    title: `Shining`,
    src: `https://cs7.pikabu.ru/post_img/2019/03/06/7/1551870696138123277.jpg`,
    description: `The Shining is a 1980 psychological horror film produced and directed by Stanley Kubrick and co-written with novelist Diane Johnson. The film is based on Stephen King's 1977 novel of the same name and stars Jack Nicholson, Shelley Duvall, Scatman Crothers, and Danny Lloyd.`,
    director: `Dude`,
    starring: `Mad Jack Nicholson`,
    genre: `Thriller`,
    date: `666`,
    background: `https://img3.goodfon.ru/wallpaper/nbig/7/2b/priroda-risovannye-oblaka.jpg`,
    ratingScore: chooseRandomInt(0, 10),
    ratingCount: chooseRandomInt(1, 450),
  }
];
