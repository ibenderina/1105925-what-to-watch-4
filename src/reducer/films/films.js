import {ALL_GENRES, APIEndpoints, CountLimit} from "@consts";
import {extend} from "@utils/utils";
import {Film} from "@api/adapter";

const ActionType = {
  SET_CURRENT_GENRE: `GET_CURRENT_GENRE`,
  GET_FILM_BY_ID: `GET_FILM_BY_ID`,
  LOAD_FILMS: `LOAD_FILMS`,
  LOAD_PROMO_FILM: `LOAD_PROMO_FILM`,
  SHOW_MORE: `SHOW_MORE`,
  TOGGLE_IF_FAVORITE: `TOGGLE_IF_FAVORITE`,
  LOAD_FAVORITE_FILMS: `LOAD_FAVORITE_FILMS`,
};

const initialState = {
  currentGenre: ALL_GENRES,
  genres: [],
  films: [],
  favoriteFilms: [],
  promoFilm: new Film({}),
  showedFilmsCount: 0,
};

const Actions = {
  loadPromoFilm: (film) => {
    return {
      type: ActionType.LOAD_PROMO_FILM,
      payload: film,
    };
  },
  loadFilms: (films) => {
    return {
      type: ActionType.LOAD_FILMS,
      payload: films,
    };
  },
  showMore: () => {
    return {
      type: ActionType.SHOW_MORE,
    };
  },
  setCurrentGenre: (genre) => {
    return {
      type: ActionType.SET_CURRENT_GENRE,
      payload: genre,
    };
  },
  toggleIsFavorite: (filmId) => {
    return {
      type: ActionType.TOGGLE_IF_FAVORITE,
      payload: filmId,
    };
  },
  loadFavoriteFilms: (filmId) => {
    return {
      type: ActionType.LOAD_FAVORITE_FILMS,
      payload: filmId,
    };
  }
};

const Operations = {
  loadPromoFilm: () => (dispatch, getState, api) => {
    return api.get(APIEndpoints.PROMO_FILM)
      .then((response) => {
        dispatch(
            Actions.loadPromoFilm(new Film(response.data))
        );
      });
  },

  loadFilms: () => (dispatch, getState, api) => {
    return api.get(APIEndpoints.FILMS)
      .then((response) => {
        dispatch(
            Actions.loadFilms(
                Film.parse(response.data)
            )
        );
      })
      .catch((err) => {
        throw err;
      });
  },

  toggleIsFavorite: (filmId, status) => (dispatch, getState, api) => {
    return api.post(`${APIEndpoints.FAVORITE}/${filmId}/${+status}`)
      .then((response) => {
        dispatch(Actions.toggleIsFavorite(new Film(response.data)));
      })
      .catch((err) => {
        throw err;
      });
  },

  loadFavoriteFilms: () => (dispatch, getState, api) => {
    return api.get(APIEndpoints.FAVORITE)
      .then((response) => {
        dispatch(Actions.loadFavoriteFilms(Film.parse(response.data)));
      })
      .catch((err) => {
        throw err;
      });
  },
};

const reducer = (state = initialState, action) => {
  switch (action.type) {
    case ActionType.SET_CURRENT_GENRE:
      return extend(state, {
        currentGenre: action.payload,
        showedFilmsCount: CountLimit.MAX_FILMS,
      });

    case ActionType.LOAD_FILMS:
      return extend(state, {
        showedFilmsCount: CountLimit.MAX_FILMS,
        films: action.payload,
      });

    case ActionType.LOAD_FAVORITE_FILMS:
      return extend(state, {
        favoriteFilms: action.payload || [],
      });

    case ActionType.LOAD_PROMO_FILM:
      return extend(state, {
        promoFilm: action.payload,
      });

    case ActionType.SHOW_MORE:
      const showedFilmsCount = state.showedFilmsCount + CountLimit.MAX_FILMS;
      return extend(state, {
        showedFilmsCount,
      });

    case ActionType.TOGGLE_IF_FAVORITE:
      const film = state.films.find((item) => item.id === action.payload.id);
      film.isFavorite = action.payload.isFavorite;
      const promoFilm = state.promoFilm.id === film.id ? action.payload : state.promoFilm;
      return extend(state, {
        films: state.films,
        promoFilm,
      });
  }

  return state;
};

export {reducer, ActionType, initialState, Actions, Operations};
