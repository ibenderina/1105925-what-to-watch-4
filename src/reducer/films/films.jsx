import {ALL_GENRES, CountLimit} from "@consts";
import {extend} from "@utils";
import {parseFilms, Film} from "@api/adapter";

const ActionType = {
  SET_CURRENT_GENRE: `GET_CURRENT_GENRE`,
  GET_FILM_BY_ID: `GET_FILM_BY_ID`,
  LOAD_FILMS: `LOAD_FILMS`,
  LOAD_PROMO_FILM: `LOAD_PROMO_FILM`,
  SHOW_MORE: `SHOW_MORE`,
};

const initialState = {
  currentGenre: ALL_GENRES,
  genres: [],
  films: [],
  isMoreFilms: true,
  promoFilm: new Film({}),
  showedFilmsCount: 0,
};

const Actions = {
  loadPromoFilm: (film) => {
    return {
      type: ActionType.LOAD_PROMO_FILM,
      payload: film,
    };
  },
  loadFilms: (films) => {
    return {
      type: ActionType.LOAD_FILMS,
      payload: films,
    };
  },
  showMore: () => {
    return {
      type: ActionType.SHOW_MORE,
    };
  },
  setCurrentGenre: (genre) => {
    return {
      type: ActionType.SET_CURRENT_GENRE,
      currentGenre: genre,
    };
  }
};

const Operations = {
  loadPromoFilm: () => (dispatch, getState, api) => {
    return api.get(`/films/promo`)
      .then((film) => {
        dispatch(
            Actions.loadPromoFilm(new Film(film.data))
        );
      });
  },
  loadFilms: () => (dispatch, getState, api) => {
    return api.get(`/films`)
      .then((films) => {
        dispatch(
            Actions.loadFilms(
                parseFilms(films.data)
            )
        );
      })
      .catch((err) => {
        throw err;
      });
  },
};

const reducer = (state = initialState, action) => {
  switch (action.type) {
    case ActionType.SET_CURRENT_GENRE:
      return extend(state, {
        currentGenre: action.currentGenre,
        showedFilmsCount: CountLimit.MAX_FILMS,
      });

    case ActionType.LOAD_FILMS:
      return extend(state, {
        showedFilmsCount: CountLimit.MAX_FILMS,
        films: action.payload,
        isMoreFilms: action.payload.length >= CountLimit.MAX_FILMS
      });

    case ActionType.LOAD_PROMO_FILM:
      return extend(state, {
        promoFilm: action.payload,
      });

    case ActionType.SHOW_MORE:
      const showedFilmsCount = state.showedFilmsCount * 2;
      return extend(state, {
        showedFilmsCount,
        isMoreFilms: state.films.length >= showedFilmsCount
      });

    case ActionType.GET_FILM_BY_ID:
      return extend(state, {
        films: action.payload,
        promoFilm: action.payload[0],
        isMoreFilms: action.payload.length >= CountLimit.MAX_FILMS
      });

    default:
      return state;
  }
};

export {reducer, ActionType, initialState, Actions, Operations};
