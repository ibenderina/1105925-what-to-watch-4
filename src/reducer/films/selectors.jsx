import {createSelector} from "reselect";
import NameSpace from "@reducer/name-space.jsx";
import {ALL_GENRES, CountLimit} from "@consts";

const getFilms = (state) => {
  return state[NameSpace.FILMS].films;
};

const getCurrentGenre = (state) => {
  return state[NameSpace.FILMS].currentGenre;
};

const getPromoFilm = (state) => {
  return state[NameSpace.FILMS].promoFilm;
};

const getShowedFilmsCount = (state) => {
  return state[NameSpace.FILMS].showedFilmsCount;
};

const getFilmId = (state, filmId) => {
  return parseInt(filmId, 10);
};

const getAllFilmsByGenre = createSelector(
    getCurrentGenre,
    getFilms,
    (currentGenre, films) => {
      if (currentGenre === ALL_GENRES) {
        return films;
      }
      return films.filter((film) => {
        return film.genre === currentGenre;
      });
    }
);


const getFilmsByGenre = createSelector(
    (_, isSimilarList) => !!isSimilarList,
    getShowedFilmsCount,
    getAllFilmsByGenre,
    (isSimilarList, showedFilmsCount, films) => {
      return films.slice(0, isSimilarList ? CountLimit.MAX_SIMILAR_FILMS : showedFilmsCount);
    }
);

const getFilmById = createSelector(
    getFilmId,
    getFilms,
    (filmId, films) => {
      if (filmId) {
        return films.find((film) => {
          return film.id === filmId;
        }) || {};
      }
      return null;
    }
);

const getGenresList = createSelector(
    getFilms,
    (films) => {
      return [...new Set(films.map((film) => {
        return film.genre;
      }))];
    }
);

const getShowMoreStatus = createSelector(
    getAllFilmsByGenre,
    getShowedFilmsCount,
    (films, showedFilmsCount) => {
      return showedFilmsCount < films.length;
    }
);

export {getFilms, getGenresList, getPromoFilm, getFilmsByGenre, getCurrentGenre, getFilmById, getShowMoreStatus};
