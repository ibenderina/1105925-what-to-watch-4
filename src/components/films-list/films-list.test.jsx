import FilmsList from "@components/films-list/films-list";
import films from "../../mocks/films";

it(`Should FilmsList render correctly`, () => {
  let tree;
  window.act(() => {
    tree = window.create(<FilmsList
      films={films}
      onClickFilmCard={() => {}}
      onFilmTitleMouseEnter={() => {}}
    />);
  });
  expect(tree.toJSON()).toMatchSnapshot();
});
