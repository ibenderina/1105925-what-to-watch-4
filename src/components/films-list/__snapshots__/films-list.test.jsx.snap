// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Should FilmsList render correctly 1`] = `
<div
  className="catalog__movies-list"
>
  <article
    className="small-movie-card catalog__movies-card"
    onClick={[Function]}
    onMouseEnter={[Function]}
  >
    <div
      className="small-movie-card__image"
    >
      <img
        alt="Dead in a Week (Or Your Money Back)"
        height="175"
        src="https://pets.mail.ru/sharepic/68/10401.png"
        width="280"
      />
    </div>
    <h3
      className="small-movie-card__title"
    >
      <a
        className="small-movie-card__link"
        href="movie-page.html"
      >
        Dead in a Week (Or Your Money Back)
      </a>
    </h3>
  </article>
  <article
    className="small-movie-card catalog__movies-card"
    onClick={[Function]}
    onMouseEnter={[Function]}
  >
    <div
      className="small-movie-card__image"
    >
      <img
        alt="The Gentlemen"
        height="175"
        src="https://twizz.ru/wp-content/uploads/-000//1/animals-about-to-drop-album-photos-58aeb0ce59ff1__700.jpg"
        width="280"
      />
    </div>
    <h3
      className="small-movie-card__title"
    >
      <a
        className="small-movie-card__link"
        href="movie-page.html"
      >
        The Gentlemen
      </a>
    </h3>
  </article>
  <article
    className="small-movie-card catalog__movies-card"
    onClick={[Function]}
    onMouseEnter={[Function]}
  >
    <div
      className="small-movie-card__image"
    >
      <img
        alt="Lion"
        height="175"
        src="https://medialeaks.ru/wp-content/uploads/2019/08/glavnaya-1.jpg"
        width="280"
      />
    </div>
    <h3
      className="small-movie-card__title"
    >
      <a
        className="small-movie-card__link"
        href="movie-page.html"
      >
        Lion
      </a>
    </h3>
  </article>
  <article
    className="small-movie-card catalog__movies-card"
    onClick={[Function]}
    onMouseEnter={[Function]}
  >
    <div
      className="small-movie-card__image"
    >
      <img
        alt="Chappie"
        height="175"
        src="https://www.thepaws.net/wp-content/uploads/2018/10/funny-chihuahua.jpg"
        width="280"
      />
    </div>
    <h3
      className="small-movie-card__title"
    >
      <a
        className="small-movie-card__link"
        href="movie-page.html"
      >
        Chappie
      </a>
    </h3>
  </article>
  <article
    className="small-movie-card catalog__movies-card"
    onClick={[Function]}
    onMouseEnter={[Function]}
  >
    <div
      className="small-movie-card__image"
    >
      <img
        alt="Guns Akimbo"
        height="175"
        src="https://memepedia.ru/wp-content/uploads/2018/12/hamster.jpg"
        width="280"
      />
    </div>
    <h3
      className="small-movie-card__title"
    >
      <a
        className="small-movie-card__link"
        href="movie-page.html"
      >
        Guns Akimbo
      </a>
    </h3>
  </article>
  <article
    className="small-movie-card catalog__movies-card"
    onClick={[Function]}
    onMouseEnter={[Function]}
  >
    <div
      className="small-movie-card__image"
    >
      <img
        alt="Undisputed"
        height="175"
        src="https://stihi.ru/pics/2013/04/04/1852.jpg"
        width="280"
      />
    </div>
    <h3
      className="small-movie-card__title"
    >
      <a
        className="small-movie-card__link"
        href="movie-page.html"
      >
        Undisputed
      </a>
    </h3>
  </article>
  <article
    className="small-movie-card catalog__movies-card"
    onClick={[Function]}
    onMouseEnter={[Function]}
  >
    <div
      className="small-movie-card__image"
    >
      <img
        alt="En man som heter Ove"
        height="175"
        src="https://techtoday.in.ua/wp-content/uploads/2018/01/GRUMPYCAT.jpg"
        width="280"
      />
    </div>
    <h3
      className="small-movie-card__title"
    >
      <a
        className="small-movie-card__link"
        href="movie-page.html"
      >
        En man som heter Ove
      </a>
    </h3>
  </article>
  <article
    className="small-movie-card catalog__movies-card"
    onClick={[Function]}
    onMouseEnter={[Function]}
  >
    <div
      className="small-movie-card__image"
    >
      <img
        alt="Shining"
        height="175"
        src="https://cs7.pikabu.ru/post_img/2019/03/06/7/1551870696138123277.jpg"
        width="280"
      />
    </div>
    <h3
      className="small-movie-card__title"
    >
      <a
        className="small-movie-card__link"
        href="movie-page.html"
      >
        Shining
      </a>
    </h3>
  </article>
</div>
`;
